syntax = "proto3";

service Spotify{
  rpc GetUser(GetUserRequest) returns (GetUserResponse) {}
  rpc GetAllUsers(Empty) returns (GetAllUsersResponse) {}
  rpc GetUserPlaylists(GetUserPlaylistsRequest) returns (GetUserPlaylistsResponse) {}
  rpc AddUser(AddUserRequest) returns (AddUserResponse) {}
  rpc UpdateUser(UpdateUserRequest) returns (UpdateUserResponse) {}
  rpc DeleteUser(DeleteUserRequest) returns (DeleteUserResponse) {}

  rpc GetPlaylist(GetPlaylistRequest) returns (GetPlaylistResponse) {}
  rpc GetAllPlaylists(Empty) returns (GetAllPlaylistsResponse) {}
  rpc GetPlaylistMusics(GetPlaylistMusicsRequest) returns (GetPlaylistMusicsResponse) {}
  rpc AddPlaylist(AddPlaylistRequest) returns (AddPlaylistResponse) {}
  rpc AddPlaylistMusic(AddPlaylistMusicRequest) returns (AddPlaylistMusicResponse) {}
  rpc UpdatePlaylist(UpdatePlaylistRequest) returns (UpdatePlaylistResponse) {}
  rpc DeletePlaylist(DeletePlaylistRequest) returns (DeletePlaylistResponse) {}

  rpc GetMusic(GetMusicRequest) returns (GetMusicResponse) {}
  rpc GetAllMusics(Empty) returns (GetAllMusicsResponse) {}
  rpc AddMusic(AddMusicRequest) returns (AddMusicResponse) {}
  rpc UpdateMusic(UpdateMusicRequest) returns (UpdateMusicResponse) {}
  rpc DeleteMusic(DeleteMusicRequest) returns (DeleteMusicResponse) {}
}

message Empty {}

message User{
  int32 id = 1;
  string name = 2;
  int32 age = 3;
  repeated int32 playlistsIds = 4;
}

message Playlist{
  int32 id = 1;
  string name = 2;
  repeated int32 musicsIds = 4;
}

message Music{
  int32 id = 1;
  string nome = 2;
  string artist = 3;
}

message GetUserRequest{
  int32 id = 1;
}

message GetUserResponse{
  User user = 1;
}

message GetAllUsersResponse{
  repeated User users = 1;
}

message GetUserPlaylistsRequest{
  int32 id = 1;
}

message GetUserPlaylistsResponse{
  repeated Playlist playlists = 1;
}

message AddUserRequest{
  User user = 1;
}

message AddUserResponse{
  int32 id = 1;
}

message UpdateUserRequest{
  User user = 1;
}

message UpdateUserResponse{
  int32 id = 1;
}

message DeleteUserRequest{
  int32 id = 1;
}

message DeleteUserResponse{
  int32 id = 1;
}

message GetPlaylistRequest{
  int32 id = 1;
}

message GetPlaylistResponse{
  Playlist playlist = 1;
}

message GetAllPlaylistsResponse{
  repeated Playlist playlists = 1;
}

message GetPlaylistMusicsRequest{
  int32 id = 1;
}

message GetPlaylistMusicsResponse{
  repeated Music musics = 1;
}

message AddPlaylistRequest{
  Playlist playlist = 1;
}

message AddPlaylistResponse{
  int32 id = 1;
}

message AddPlaylistMusicRequest{
  int32 playlistId = 1;
  int32 musicId = 2;
}

message AddPlaylistMusicResponse{
  int32 id = 1;
}

message UpdatePlaylistRequest{
  Playlist playlist = 1;
}

message UpdatePlaylistResponse{
  int32 id = 1;
}

message DeletePlaylistRequest{
  int32 id = 1;
}

message DeletePlaylistResponse{
  int32 id = 1;
}

message GetMusicRequest{
  int32 id = 1;
}

message GetMusicResponse{
  Music music = 1;
}

message GetAllMusicsResponse{
  repeated Music musics = 1;
}

message AddMusicRequest{
  Music music = 1;
}

message AddMusicResponse{
  int32 id = 1;
}

message UpdateMusicRequest{
  Music music = 1;
}

message UpdateMusicResponse{
  int32 id = 1;
}

message DeleteMusicRequest{
  int32 id = 1;
}

message DeleteMusicResponse{
  int32 id = 1;
}